pantheon-bundle/src/test/java/com/redhat/pantheon/model/api/SlingModelsTest.java

```diff
@@ -0,0 +1,315 @@
+        Grandchild grandchild = child.createChild("grandchild", Grandchild.class);
+
+        // Then
+        assertNotNull(child);
+        assertNotNull(grandchild);
+        assertThrows(RuntimeException.class, () -> model.createChild("child", ChildResource.class),
+                "Same child cannot be created twice");
+    }
+
+    @Test
+    void getChild() {
+        // Given
+        sc.build()
+                .resource("/test/child/grandchild")
+                .commit();
+
+        // When
+        TestResource model = SlingModels.getModel(sc.resourceResolver(), "/test", TestResource.class);
+        ChildResource child = model.getChild("child", ChildResource.class);
+        Grandchild grandchild = child.getChild("grandchild", Grandchild.class);
+
+        // Then
+        assertNotNull(child);
+        assertNotNull(grandchild);
+        assertNull(model.getChild("nonexistent"));
+    }
+
+    @Test
+    void getOrCreateChild() {
+        // Given
+        sc.build()
+                .resource("/test/child")
+                .commit();
+
+        // When
+        TestResource model = SlingModels.getModel(sc.resourceResolver(), "/test", TestResource.class);
+
+        // Then
+        assertNotNull(model.child().get());
+        assertEquals(model.child().get().getName(), model.child().getOrCreate().getName());
+        assertNull(model.child().get().grandchild().get());
+        assertNotNull(model.child().get().grandchild().getOrCreate());
+        assertNotNull(model.child().get().grandchild().get());
+    }
+
+    @Test
+    void create() throws Exception {
+        // Given
+        ResourceResolver resourceResolver = sc.resourceResolver();
+        Resource root = resourceResolver.getResource("/");
+
+        // When
+        TestResource model = SlingModels.getModel(resourceResolver.create(root,"test", null), TestResource.class);
+        model.name().set("new name");
+        model.longField().set(10L);
+        resourceResolver.commit();
+
+        // Then
+        TestResource storedModel = SlingModels.getModel(resourceResolver, "/test", TestResource.class);
+        assertEquals("new name", storedModel.name().get());
+        assertEquals(new Long(10), storedModel.longField().get());
+    }
+
+    @Test
+    void toMap() {
+        // Given
+        String[] arrayValue = {"A", "B", "C"};
+        sc.build()
+                .resource("/test",
+                        "name", "new name",
+                        "longField", 20L,
+                        "intField", 10,
+                        "booleanField", true,
+                        "dateField", Calendar.getInstance(),
+                        "stringArrayField", arrayValue,
+                        "enumField", TestResource.Value.VALUE_2.name())
+                .commit();
+
+        // When
+        Map<String, Object> map = SlingModels.getModel(sc.resourceResolver(), "/test", TestResource.class)
+                .getValueMap();
```

Comment:
This is quite the test just for 1 line of code, which AFAICT is just the pass-through method `ResourceDecorator.getValueMap()`... is there more to it than that?

Reply:
Yes, so the real test is that when the model is built (via the `getModel` call), all different types of fields are mapped correctly. If you look at the assertions below, we are testing the correct mapping for fields of type: stirng, long, date, boolean, string array, and enum.

Reply:
I see the assertions, but none of our code is doing the mapping, correct?

Reply:
Yes it is. Specifically the `FieldImpl` class.

Reply:
Tell me how, then? You are calling `getValueMap()`. That is provided by the `org.apache.sling.api.resource.Resource` package. I could *imagine* that maybe the valueMap gets populated via methods in `FieldImpl`, but when I look at that class, I don't see that as likely - especially because `FieldImpl.get()` *calls* `getValueMap()` *itself* in order to function.

It would be one thing if your test assertions were using `SlingModel.get()` calls. But as long as they're using `valueMap.get()` calls, then I don't see our code being evaluated there.

Reply:
You are correct. I was thinking of the `getters()` test method and not this one for some reason.

So right now the `toMap` method is being delegated to the resource class, and the expectation is that the fields will appear in the resulting map as described by the test. IF we ever decide to change that and say, have our own custom implementation of `toMap`, then this would serve as a regression to keep the expectations of the generated map.

Reply:
Ok... that's a fair point. Although this doesn't test our code today, it doesn't hurt to have it to future-proof things. Thanks for explaining.